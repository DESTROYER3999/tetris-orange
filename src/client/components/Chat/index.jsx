import { connect } from "react-redux";
import ChatSub from "./subcomponent";
import { newChatMessage } from "../../actions/chat";

const mapStateToProps = state => {
  return {
    chatMessages: state.chatMessages,
    players: state.players
  };
};

const mapDispatchToProps = dispatch => {
  return {
    onSubmit: message => {
      dispatch(newChatMessage(message));
    }
  };
};

const Chat = connect(
  mapStateToProps,
  mapDispatchToProps
)(ChatSub);

export default Chat;
