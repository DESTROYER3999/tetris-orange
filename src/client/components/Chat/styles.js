import styled from "styled-components";

export const StyledMainContainer = styled.div`
  display: flex;
  flex: 4;
  padding: 10px;
  margin: 5px;
  background-color: rgba(0, 0, 0, 0.3);
  border-radius: 2px;
`;

export const StyledChatContainer = styled.div`
  display: flex;
  flex: 1;
  border: 3px solid white;
  border-radius: 5px;
  height: 0;
  min-height: 100%;
`;

export const StyledLeftContainer = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;
  background-color: rgba(0, 0, 0, 0.2);
  border-right: 3px solid white;
`;

export const StyledRightContainer = styled.div`
  flex: 4;
  display: flex;
  flex-direction: column;
`;

export const StyledTitleContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 4vh;
  border-bottom: 3px solid white;
`;
export const StyledMessagesContainer = styled.div`
  display: block;
  flex: 1;
  padding: 5px;
  flex-wrap: wrap;
  overflow-y: auto;
  overflow-x: hidden;
  text-overflow: ellipsis;
`;

export const StyledForm = styled.form`
  display: flex;
  padding: 5px;
`;

export const StyledPlayerListContainer = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;
  align-items: center;
  padding: 5px;
  border-bottom: 5px;
  overflow-y: auto;
  overflow-x: hidden;
  text-overflow: ellipsis;
`;

export const StyledText = styled.div`
  font-family: Verdana;
  font-size: 1.5vh;
  min-width: 100%;
  width: 0;
  & font {
    font-family: Verdana;
    font-weight: 600;
  }
`;

export const StyledInput = styled.input`
  height: 4vh;
  padding: 0.5vh;
  padding-left: 1vh;
  border-radius: 5px;
  border: 2px solid white;
  font-size: 2vh;
  width: 100%;
  background: none;
  outline: none;
  color: white;
  font-family: Verdana;
  font-size: 1.5vh;
  ::placeholder {
    color: white;
  }
`;
