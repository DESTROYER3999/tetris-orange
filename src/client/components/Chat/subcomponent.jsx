import React, { useState, useEffect, useRef } from "react";
import {
  StyledMainContainer,
  StyledChatContainer,
  StyledLeftContainer,
  StyledRightContainer,
  StyledTitleContainer,
  StyledMessagesContainer,
  StyledForm,
  StyledPlayerListContainer,
  StyledInput,
  StyledText
} from "./styles";
import { handleKeyPress } from "../../actions/game";
import PropTypes from "prop-types";

const ChatSub = ({ chatMessages, players, onSubmit }) => {
  const [message, setMessage] = useState("");

  const handleChange = e => {
    const value = e.target.value;
    setMessage(value);
  };

  const handleSubmit = e => {
    e.preventDefault();
    onSubmit(message);
    setMessage("");
  };

  const disableKeyPressListener = () => {
    document.removeEventListener("keydown", handleKeyPress);
  };

  const enableKeyPressListener = () => {
    document.addEventListener("keydown", handleKeyPress);
  };

  const messagesEndRef = useRef(null);

  const scrollToBottom = () => {
    messagesEndRef.current.scrollIntoView({ behavior: "smooth" });
  };

  useEffect(scrollToBottom, [chatMessages]);

  const displayChatMessages = () =>
    chatMessages.map((message, index) => {
      if (message.type === "message") {
        return (
          <StyledText key={index}>
            <font>{message.author}:</font> {message.text}
          </StyledText>
        );
      } else {
        return (
          <StyledText key={index}>
            <font style={{ color: message.color }}>{message.text}</font>
          </StyledText>
        );
      }
    });

  return (
    <StyledMainContainer id="chat">
      <StyledChatContainer>
        <StyledLeftContainer>
          <StyledTitleContainer>
            <p>Players</p>
          </StyledTitleContainer>
          <StyledPlayerListContainer>
            {players.map((player, index) => (
              <StyledText key={index}>{player}</StyledText>
            ))}
          </StyledPlayerListContainer>
        </StyledLeftContainer>
        <StyledRightContainer>
          <StyledTitleContainer>
            <p>Messages</p>
          </StyledTitleContainer>
          <StyledMessagesContainer>
            {displayChatMessages()}
            <div ref={messagesEndRef} />
          </StyledMessagesContainer>
          <StyledForm onSubmit={e => handleSubmit(e)}>
            <StyledInput
              type="text"
              id="chatMessage"
              spellCheck="false"
              autoComplete="off"
              placeholder="Write a message ..."
              value={message}
              onChange={e => handleChange(e)}
              onFocus={() => disableKeyPressListener()}
              onBlur={() => enableKeyPressListener()}
            />
          </StyledForm>
        </StyledRightContainer>
      </StyledChatContainer>
    </StyledMainContainer>
  );
};

ChatSub.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  chatMessages: PropTypes.array.isRequired,
  players: PropTypes.array.isRequired
};

export default ChatSub;
