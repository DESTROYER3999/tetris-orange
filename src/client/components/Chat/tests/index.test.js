import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import configureStore from "redux-mock-store";
import Chat from "../index";
import socket from "../../../services/socket-api";
import thunk from "redux-thunk";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {
  chatMessages: [],
  players: []
};

const store = mockStore(initialState);

beforeEach(done => {
  // Setup
  if (socket.disconnected) {
    socket.connect();
  }
  socket.on("connect", () => {
    done();
  });
});

afterEach(done => {
  //   Cleanup;
  if (socket.connected) {
    socket.disconnect();
  }
  done();
});

describe("<Chat />", () => {
  test("Chat rendering with store without crashing", () => {
    const wrapper = shallow(<Chat store={store} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
