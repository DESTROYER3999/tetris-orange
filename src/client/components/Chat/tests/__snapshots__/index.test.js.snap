// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chat /> Chat rendering with store without crashing 1`] = `
<ContextProvider
  value={null}
>
  <ChatSub
    chatMessages={Array []}
    onSubmit={[Function]}
    players={Array []}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;
